### PR TITLE
NCC-E003660-WHE: Low Entropy Bootstrap Tokens

### LINKED ISSUE #119639: NCC-E003660-WHE: Low Entropy Bootstrap Tokens
### NCC-E003660-WHE: Low Entropy Bootstrap Tokens
This issue was reported in the [Kubernetes 1.24 Security Audit Report](https://github.com/kubernetes/sig-security/blob/main/sig-security-external-audit/security-audit-2021-2022/findings/Kubernetes%20v1.24%20Final%20Report.pdf)

**Impact**
The level of entropy in bootstrap token secrets does not meet security best practice. If an
attacker is able to acquire a bootstrap token, they could obtain cluster admin privileges by
joining their own node to the cluster as a master node.

**Description**
The bootstrap token secrets used by the Kubernetes API server consist of 16 base-36
characters, which provides approximately 83 bits of entropy. It is conventional to use at
least 128 bits of entropy for static secrets or symmetric keys, and indeed the naming of the
`BootstrapTokenSecretBytes` variable in the source code (which takes the value 16)
suggests that this was perhaps the intention here. Note that it is possible to perform an
offline brute-force attack to recover bootstrap token secrets (based on the signature of the
`cluster-info` configmap), and while this is not realistically exploitable with current
technology, the length of the token secrets should be increased in line with security best
practice.

**Recommendation**
Increase the length of bootstrap token secrets so that they include at least 128 bits of
entropy.

**Component**
kube-apiserver

**Location**
Function `GenerateBootstrapToken()` in file [helpers.go:L45](https://github.com/kubernetes/kubernetes/blob/v1.24.0/staging/src/k8s.io/cluster-bootstrap/token/util/helpers.go#L45)

**Anything else we need to know?**
See umbrella issue [#118980](https://github.com/kubernetes/kubernetes/issues/118980) for current status of all issues created from these findings.
The vendor gave this issue an ID of NCC-E003660-WHE and it was finding 17 of the report under the kube-apiserver section.
The vendor considers this issue Risk as Informational, High Impact, Exploitability at None
To view the original finding, begin on page 47 of the [Kubernetes 1.24 Security Audit Report](https://github.com/kubernetes/sig-security/blob/main/sig-security-external-audit/security-audit-2021-2022/findings/Kubernetes%20v1.24%20Final%20Report.pdf)

**Test Environment**
Kubernetes 1.24.3

### PR DESCRIPTION
**Pull request by charles-chenzz**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
follow the suggestion from #119639 
#### Which issue(s) this PR fixes:

Fixes #119639 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by charles-chenzz** [comment]
don't know why the e2e-kind failed, might need some help

File: staging/src/k8s.io/cluster-bootstrap/token/api/types.go [hunk 1]
```diff
@@ -96,7 +96,7 @@ const (
 	BootstrapGroupPattern = `\Asystem:bootstrappers:[a-z0-9:-]{0,255}[a-z0-9]\z`
 
 	// BootstrapTokenPattern defines the {id}.{secret} regular expression pattern
-	BootstrapTokenPattern = `\A([a-z0-9]{6})\.([a-z0-9]{16})\z`
+	BootstrapTokenPattern = `\A([a-z0-9]{6})\.([a-z0-9]{24})\z`
```
Comment (liggitt) [inline staging/src/k8s.io/cluster-bootstrap/token/api/types.go:99]: this is an incompatible change that will break all callers of ParseToken / IsValidBootstrapToken / etc

Comment (charles-chenzz) [reply staging/src/k8s.io/cluster-bootstrap/token/api/types.go:99]: if so, do we still want to fix the low entropy bootstrap tokens problem?

**Comment by ibihim** [comment]
This needs a KEP: https://github.com/kubernetes/kubernetes/issues/119639#issuecomment-1678555859

**Pull request closed without merging**